doc/how_to_guides/planner_cost_functions/src/planner_cost_functions_main.cpp

```diff
@@ -0,0 +1,355 @@
+    }
+    visual_tools_.prompt("Press 'next' in the RvizVisualToolsGui window to start the demo");
+    visual_tools_.trigger();
+    return true;
+  }
+
+  struct PipelineConfig
+  {
+    std::string planning_pipeline;
+    std::string planner_id;
+    bool use_cost_function;
+  };
+
+  /// \brief Request a motion plan based on the assumption that a goal is set and print debug information.
+  void planAndVisualize(std::vector<PipelineConfig> pipeline_configs,
+                        moveit_msgs::msg::MotionPlanRequest const& motion_plan_request)
+  {
+    visual_tools_.deleteAllMarkers();
+    visual_tools_.prompt("Press 'next' in the RvizVisualToolsGui window to continue the demo");
+
+    // Set goal state
+    planning_component_->setGoal(motion_plan_request.goal_constraints);
+    // Set start state as current state
+    planning_component_->setStartStateToCurrentState();
+
+    // Get start state
+    auto robot_start_state = planning_component_->getStartState();
+
+    // Get planning scene
+    auto planning_scene_monitor = moveit_cpp_->getPlanningSceneMonitorNonConst();
+    planning_scene_monitor->updateFrameTransforms();
+    auto planning_scene = [planning_scene_monitor] {
+      planning_scene_monitor::LockedPlanningSceneRO ls(planning_scene_monitor);
+      return planning_scene::PlanningScene::clone(ls);
+    }();
+
+    auto group_name = motion_plan_request.group_name;
+
+    moveit_cpp::PlanningComponent::PlanRequestParameters plan_request_parameters;
+    plan_request_parameters.load(node_);
+    RCLCPP_INFO_STREAM(
+        LOGGER, "Default plan request parameters loaded with --"
+                    << " planning_pipeline: " << plan_request_parameters.planning_pipeline << ','
+                    << " planner_id: " << plan_request_parameters.planner_id << ','
+                    << " planning_time: " << plan_request_parameters.planning_time << ','
+                    << " planning_attempts: " << plan_request_parameters.planning_attempts << ','
+                    << " max_velocity_scaling_factor: " << plan_request_parameters.max_velocity_scaling_factor << ','
+                    << " max_acceleration_scaling_factor: " << plan_request_parameters.max_acceleration_scaling_factor);
+
+    std::vector<planning_interface::MotionPlanResponse> solutions;
+    solutions.reserve(pipeline_configs.size());
+    for (auto const& pipeline_config : pipeline_configs)
+    {
+      plan_request_parameters.planning_pipeline = pipeline_config.planning_pipeline;
+      plan_request_parameters.planner_id = pipeline_config.planner_id;
+
+      // Set cost function
+      if (pipeline_config.use_cost_function)
+      {
+        planning_component_->setStateCostFunction(
+            [robot_start_state, group_name, planning_scene](const Eigen::VectorXd& state_vector) mutable {
+              auto clearance_cost_fn =
+                  moveit::cost_functions::getMinJointDisplacementCostFn(*robot_start_state, group_name, planning_scene);
```

Comment:
Related to my comment in https://github.com/ros-planning/moveit2/pull/2153#discussion_r1371384918, here is where I think you could do:

```
auto clearance_cost_fn =
                  std::bind(moveit::cost_functions::computeMinJointDisplacementCostFn(*robot_start_state, group_name, planning_scene, std::placeholders::_1));
```

The advantage is that `cost_functions` wouldn't have to depend on the planning interface StateCostFn, and the whole process is more transparent: you have a function that computes a cost, and you bind it here into get a StateCostFn.